src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -780,6 +784,79 @@ public void MarkSymbolEndAnalysisComplete(ISymbol symbol, DiagnosticAnalyzer ana
             diagReporter.Free();
         }
 
+        /// <summary>
+        /// Tries to execute the additional file actions.
+        /// </summary>
+        /// <param name="additionalFileActions">Actions to be executed.</param>
+        /// <param name="analyzer">Analyzer whose actions are to be executed.</param>
+        /// <param name="file">Additional file to analyze.</param>
+        /// <param name="analysisScope">Scope for analyzer execution.</param>
+        /// <param name="analysisStateOpt">An optional object to track analysis state.</param>
+        /// <returns>
+        /// True, if successfully executed the actions for the given analysis scope OR all the actions have already been executed for the given analysis scope.
+        /// False, if there are some pending actions that are currently being executed on another thread.
+        /// </returns>
+        public bool TryExecuteAdditionalFileActions(
+            ImmutableArray<AdditionalFileAnalyzerAction> additionalFileActions,
+            DiagnosticAnalyzer analyzer,
+            SourceOrNonSourceFile file,
+            AnalysisScope analysisScope,
+            AnalysisState analysisStateOpt)
+        {
+            RoslynDebug.Assert(file.NonSourceFile != null);
+            AnalyzerStateData analyzerStateOpt = null;
+
+            try
+            {
+                if (TryStartSyntaxAnalysis(file, analyzer, analysisScope, analysisStateOpt, out analyzerStateOpt))
+                {
+                    ExecuteAdditionalFileActionsCore(additionalFileActions, analyzer, file, analyzerStateOpt);
+                    analysisStateOpt?.MarkSyntaxAnalysisComplete(file, analyzer);
+                    return true;
+                }
+
+                return analysisStateOpt == null || !analysisStateOpt.HasPendingSyntaxAnalysis(analysisScope);
+            }
+            finally
+            {
+                analyzerStateOpt?.ResetToReadyState();
+            }
+        }
+
+        private void ExecuteAdditionalFileActionsCore(
+            ImmutableArray<AdditionalFileAnalyzerAction> additionalFileActions,
+            DiagnosticAnalyzer analyzer,
+            SourceOrNonSourceFile file,
+            AnalyzerStateData analyzerStateOpt)
+        {
+            RoslynDebug.Assert(file.NonSourceFile != null);
+            var additionalFile = file.NonSourceFile;
+
+            var diagReporter = GetAddSyntaxDiagnostic(file, analyzer);
+
+            using var _ = PooledDelegates.GetPooledFunction((d, arg) => arg.self.IsSupportedDiagnostic(arg.analyzer, d), (self: this, analyzer), out Func<Diagnostic, bool> isSupportedDiagnostic);
```

Comment:
>PooledDelegates.GetPooledFunction [](start = 26, length = 33)

Is there a reason to pool this delegate? It looks like it doesn't capture any state so it will be stored in a static field.